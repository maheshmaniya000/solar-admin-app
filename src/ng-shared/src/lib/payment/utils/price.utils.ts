import { Product } from '@solargis/types/customer';
import { ProductVariant } from '@solargis/types/order-invoice';
import { sum } from '@solargis/types/utils';

import { ProductMap } from '../payment.types';

/**
 * Calculate base product price
 *
 * @param productVariant
 * @param product
 */
 export function getProductBasePrice(productVariant: ProductVariant, product: Product): number {
  return typeof product.price === 'string' ? parseInt(product.price, 10) : product.price;
}

/**
 * Calculate additional price per user
 *
 * @param productVariant
 * @param product
 */
 export function getProductUsersPrice(productVariant: ProductVariant, product: Product): number {
  if (productVariant.totalUsers) {
    const freeUsers = product.techSpec.autoProcessDefinition.usersLimit;
    return Math.max(productVariant.totalUsers - freeUsers, 0) * product.techSpec.additionalUserCost;
  } else {
    return 0;
  }
}

/**
 * Calculate price for single product
 *
 * @param productVariant
 * @param product
 */
 export function getproductVariantPrice(productVariant: ProductVariant, product: Product): number {
  return getProductBasePrice(productVariant, product) + getProductUsersPrice(productVariant, product);
}

/**
 * Calculate price for multiple products
 *
 * @param productVariants
 * @param products
 */
export function getProductVariantsPrice(productVariants: ProductVariant[], products: ProductMap): number {
  return sum(productVariants.map(p => getproductVariantPrice(p, products[p.code])));
}

