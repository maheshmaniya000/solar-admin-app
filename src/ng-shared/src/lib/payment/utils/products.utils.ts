import { Product } from '@solargis/types/customer';
import { ProspectProductCodes } from '@solargis/types/order-invoice';


export function getProductNameTranslation(product: Product): string {
  const map = {
    [ProspectProductCodes.PROSPECT_BASIC]: 'PROSPECT_BASIC',
    [ProspectProductCodes.PROSPECT_PRO]: 'PROSPECT_PRO',
    [ProspectProductCodes.BASIC_TO_PRO]: 'UPGRADE',
  };

  if (map[product.code]) {
    return 'common.subscriptionType.' + map[product.code];
  } else {
    return product.descriptionENG;
  }
}
