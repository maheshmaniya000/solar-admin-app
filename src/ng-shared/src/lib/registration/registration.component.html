<button
  class="sg-small-button"
  mat-flat-button
  color="accent"
  (click)="onOpenRegistrationButtonClick()"
  data-test="open-registration-button"
>
  {{ 'auth.registration.title' | transloco }}
</button>

<ng-template #registrationDialogContent>
  <ng-container
    *ngIf="registrationStore.registrationDialogViewModel$ | async as viewModel"
  >
    <sg-dialog
      *transloco="let t; read: 'auth.registration'"
      [title]="t('createUserAccount')"
      [closeIconButtonEnabled]="true"
      [dividedContent]="false"
    >
      <mat-horizontal-stepper
        linear
        class="no-header"
        [selectedIndex]="viewModel.step"
      >
        <mat-step> step1 </mat-step>
        <mat-step> step2 </mat-step>
        <mat-step> step3 </mat-step>
        <mat-step> step4 </mat-step>
        <mat-step> step5 </mat-step>
        <mat-step> step6 </mat-step>
      </mat-horizontal-stepper>

      <mat-dialog-actions>
        <button
          *ngIf="viewModel.backButtonDisplayed"
          mat-button
          color="primary"
          (click)="registrationStore.setPreviousStep()"
          data-test="back-button"
        >
          <mat-icon>arrow_back</mat-icon>
          {{ 'common.action.back' | transloco }}
        </button>
        <span class="spacer"></span>
        <button
          class="sg-large-button"
          *ngIf="viewModel.continueButtonDisplayed"
          mat-flat-button
          color="accent"
          (click)="registrationStore.setNextStep()"
          data-test="continue-button"
        >
          {{ 'common.action.continue' | transloco }}
        </button>
      </mat-dialog-actions>

      <sg-dialog-footer *ngIf="viewModel.footerDisplayed">
        {{ t('alreadyHaveAccount') }}
        <button
          mat-button
          color="accent"
          (click)="onLogInButtonClick()"
          data-test="login-button"
        >
          {{ t('login') }}
        </button>
      </sg-dialog-footer>
    </sg-dialog>
  </ng-container>
</ng-template>
