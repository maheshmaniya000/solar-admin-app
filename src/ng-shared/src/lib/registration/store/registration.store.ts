import { Injectable } from '@angular/core';
import { ComponentStore } from '@ngrx/component-store';
import { Observable } from 'rxjs';

import { AccountType } from '../models/account-type.enum';
import { RegistrationStep } from '../models/registration-step.enum';

type RegistrationState = {
  accountType: AccountType;
  step: RegistrationStep;
};

type RegistrationViewModel = {
  step: RegistrationStep;
  backButtonDisplayed: boolean;
  continueButtonDisplayed: boolean;
  footerDisplayed: boolean;
};

const initialState: RegistrationState = {
  step: RegistrationStep.selectAccountType,
  accountType: AccountType.work
};

@Injectable()
export class RegistrationStore extends ComponentStore<RegistrationState> {
  readonly setPreviousStep = this.updater(state => ({
    ...state,
    step: --state.step
  }));

  readonly setNextStep = this.updater(state => ({
    ...state,
    step: ++state.step
  }));

  readonly setAccountType = this.updater((state, accountType: AccountType) => ({
    ...state,
    accountType
  }));

  readonly registrationDialogViewModel$: Observable<RegistrationViewModel> =
    this.select(state => ({
      ...state,
      backButtonDisplayed: state.step !== RegistrationStep.selectAccountType,
      continueButtonDisplayed:
        state.step !== RegistrationStep.accountVerification,
      footerDisplayed: state.step === RegistrationStep.selectAccountType
    }));

  constructor() {
    super(initialState);
  }
}
