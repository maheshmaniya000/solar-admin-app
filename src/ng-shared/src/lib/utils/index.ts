export * from './router/custom.reuse-stragegy';
export * from './router/custom.router-state.serializer';
export * from './router/custom.url-serializer';
export * from './rxjs/simplechange.operator';
export * from './app-device.utils';
export * from './dom.utils';
export * from './enum.utils';
export * from './misc';
export * from './module.utils';
export * from './response';
export * from './timezone.utils';
export * from './translation.utils';
export * from './validators';
