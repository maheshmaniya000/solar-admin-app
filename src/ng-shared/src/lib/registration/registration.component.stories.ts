import { APP_BASE_HREF } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { RouterTestingModule } from '@angular/router/testing';
import { provideMockStore } from '@ngrx/store/testing';
import { moduleMetadata } from '@storybook/angular';

import { SgDefaultsModule } from 'components/defaults/defaults.module';

import { TranslocoRootModule } from '../core/transloco-root.module';
import { RegistrationModule } from './registration.module';

export default {
  title: 'Solargis/Components/Registration',
  decorators: [
    moduleMetadata({
      imports: [
        BrowserAnimationsModule,
        SgDefaultsModule,
        RouterTestingModule.withRoutes([]),
        HttpClientModule,
        TranslocoRootModule,
        RegistrationModule
      ],
      providers: [{ provide: APP_BASE_HREF, useValue: '.' }, provideMockStore()]
    })
  ]
};

export const basic = (props): any => ({
  props,
  template: `
    <sg-registration></sg-registration>
  `
});
