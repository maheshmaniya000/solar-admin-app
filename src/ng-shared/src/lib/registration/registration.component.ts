import {
  ChangeDetectionStrategy,
  Component,
  TemplateRef,
  ViewChild
} from '@angular/core';
import { Store } from '@ngrx/store';

import { DialogService } from 'components/dialog/service/dialog.service';

import { RequireUserLogin } from '../user/actions/auth.actions';
import { RegistrationStep } from './models/registration-step.enum';
import { RegistrationStore } from './store/registration.store';

@Component({
  selector: 'sg-registration',
  templateUrl: './registration.component.html',
  providers: [RegistrationStore],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class RegistrationComponent {
  @ViewChild('registrationDialogContent')
  registrationDialogContent: TemplateRef<any>;

  readonly registrationStep = RegistrationStep;

  constructor(
    readonly registrationStore: RegistrationStore,
    private readonly dialogService: DialogService,
    private readonly store: Store
  ) {}

  onOpenRegistrationButtonClick(): void {
    this.dialogService.openLargeDialog(this.registrationDialogContent, {
      disableClose: true
    });
  }

  onLogInButtonClick(): void {
    this.store.dispatch(new RequireUserLogin(undefined));
  }
}
