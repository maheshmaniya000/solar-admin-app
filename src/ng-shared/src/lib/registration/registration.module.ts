import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { MatButtonModule } from '@angular/material/button';
import { MatIconModule } from '@angular/material/icon';
import { TranslocoModule } from '@ngneat/transloco';

import { DialogModule } from 'components/dialog/dialog.module';
import { StepperModule } from 'components/stepper/stepper.module';

import { RegistrationComponent } from './registration.component';

@NgModule({
  imports: [
    CommonModule,
    MatButtonModule,
    MatIconModule,
    TranslocoModule,
    DialogModule,
    StepperModule
  ],
  declarations: [RegistrationComponent],
  exports: [RegistrationComponent]
})
export class RegistrationModule {}
