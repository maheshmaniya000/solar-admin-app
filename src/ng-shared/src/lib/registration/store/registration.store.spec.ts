import { createServiceFactory, SpectatorService } from '@ngneat/spectator/jest';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';

import { runMarbleTest } from 'components/utils/test';

import { AccountType } from '../models/account-type.enum';
import { RegistrationStep } from '../models/registration-step.enum';
import { RegistrationStore } from './registration.store';

describe('RegistrationStore', () => {
  const initialState = {
    step: RegistrationStep.selectAccountType,
    accountType: AccountType.work
  };
  let spectator: SpectatorService<RegistrationStore>;
  const createService = createServiceFactory({
    service: RegistrationStore
  });

  const getStep$ = (store: RegistrationStore): Observable<RegistrationStep> =>
    store.state$.pipe(map(({ step }) => step));

  beforeEach(() => {
    spectator = createService();
  });

  it('should match snapshot of enum with correct order of values', () => {
    expect(RegistrationStep).toMatchSnapshot();
  });

  describe('initial state', () => {
    it('should emit correct initial state', () => {
      runMarbleTest(spectator.service.state$).andExpectToEmit('a-', {
        a: initialState
      });
    });
  });

  describe('setPreviousStep', () => {
    it('should emit state with previous step', () => {
      spectator.service.patchState({
        step: RegistrationStep.accountVerification
      });
      spectator.service.setPreviousStep();

      runMarbleTest(getStep$(spectator.service)).andExpectToEmit('a-', {
        a: RegistrationStep.termsAndConditions
      });
    });
  });

  describe('setNextStep', () => {
    it('should emit state with next step', () => {
      spectator.service.setNextStep();

      runMarbleTest(getStep$(spectator.service)).andExpectToEmit('a-', {
        a: RegistrationStep.userInformation
      });
    });
  });

  describe('registrationDialogViewModel$', () => {
    const testProperty = (
      propertyName:
        | 'backButtonDisplayed'
        | 'continueButtonDisplayed'
        | 'footerDisplayed',
      testData: { step: RegistrationStep; expectedValue: boolean }[]
    ): void =>
      testData.forEach(({ step, expectedValue }) =>
        it(`should emit '${expectedValue}' when step is '${RegistrationStep[step]}'`, () => {
          spectator.service.patchState({
            step
          });

          runMarbleTest(
            spectator.service.registrationDialogViewModel$.pipe(
              map(state => state[propertyName])
            )
          ).andExpectToEmit('a-', {
            a: expectedValue
          });
        })
      );

    describe('backButtonDisplayed', () => {
      testProperty('backButtonDisplayed', [
        {
          step: RegistrationStep.selectAccountType,
          expectedValue: false
        },
        {
          step: RegistrationStep.studentInformation,
          expectedValue: true
        }
      ]);
    });

    describe('continueButtonDisplayed', () => {
      testProperty('continueButtonDisplayed', [
        {
          step: RegistrationStep.accountVerification,
          expectedValue: false
        },
        {
          step: RegistrationStep.studentInformation,
          expectedValue: true
        }
      ]);
    });

    describe('footerDisplayed', () => {
      testProperty('footerDisplayed', [
        {
          step: RegistrationStep.accountVerification,
          expectedValue: false
        },
        {
          step: RegistrationStep.selectAccountType,
          expectedValue: true
        }
      ]);
    });
  });
});
