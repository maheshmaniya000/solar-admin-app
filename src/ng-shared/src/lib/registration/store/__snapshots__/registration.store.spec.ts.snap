// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegistrationStore should match snapshot of enum with correct order of values 1`] = `
Object {
  "0": "selectAccountType",
  "1": "userInformation",
  "2": "studentInformation",
  "3": "trialSelection",
  "4": "termsAndConditions",
  "5": "accountVerification",
  "accountVerification": 5,
  "selectAccountType": 0,
  "studentInformation": 2,
  "termsAndConditions": 4,
  "trialSelection": 3,
  "userInformation": 1,
}
`;
